File: target_practice_solutions/go/w4_d1_oop.go
package main

import (
	"fmt"
)

type Board struct {
  matrix [][]rune
  symbol []rune
  numRows, numCols int
}


func (board *Board) Init(numRows int, numCols int) *Board {
  board.matrix = make([][]rune, numRows)
  for rowInd := range board.matrix {
    board.matrix[rowInd] = make([]rune, numCols)
    for colInd := range board.matrix[rowInd] {
      board.matrix[rowInd][colInd] = '-'
    }
  }
  board.symbol = []rune{'O', 'X'}
  board.numRows = numRows
  board.numCols = numCols
  return board
}


func (board *Board) printBoard() {
    output := ""
    for rowInd := 0; rowInd < board.numRows; rowInd++ {
        output += "|"
        for colInd := 0; colInd < board.numCols; colInd++ {
            output += " "
            output += string(board.matrix[rowInd][colInd])
            output += " "
            output += "|"
        }
        output += "\n"
    }
    fmt.Println(output)
}

func (board *Board) canPlacePiece(row int, col int) bool {
    return (row < board.numRows && row >= 0) &&
           (col < board.numCols && col >= 0) &&
           board.matrix[row][col]== '-'
}

func (board *Board) placePiece(row int, col int, player int) {
    if (board.canPlacePiece(row, col)) {
        board.matrix[row][col] = board.symbol[player]
    }
}

func (board *Board) checkWinCondtion(player int) bool {
    return board.checkRows(player) || board.checkColumns(player) || board.checkDiagonals(player)
}

func (board *Board) checkDiagonals(player int) bool {
    hasWin := true
    for i := 0; i < board.numRows; i++ {
        if board.matrix[i][i]== '-' || board.matrix[i][i] != board.symbol[player] {
            hasWin = false
            break
        }
    }
    if hasWin { return true }

    hasWin = true
    // Check other diagonal
    for i := 0; i < board.numRows; i++ {
        if board.matrix[i][board.numRows-1 - i]== '-' || board.matrix[i][board.numRows-1 - i] != board.symbol[player] {
            hasWin = false
            break
        }
    }
    return hasWin
}

func (board *Board) checkRows(player int) bool {
    var hasWin bool
    for row :=0; row < board.numRows; row++ {
        hasWin = true
        for  col :=0; col < board.numCols; col++ {
            if board.matrix[row][col]== '-' || board.matrix[row][col] != board.symbol[player] {
                hasWin = false
                break
            }
        }
        if hasWin { return true }
    }
    return false
}

func (board *Board) checkColumns(player int) bool {
    var hasWin bool
    for  col :=0; col < board.numCols; col++ {
        hasWin = true
        for row :=0; row < board.numRows; row++ {
            if board.matrix[row][col]== '-' || board.matrix[row][col] != board.symbol[player] {
                hasWin = false
                break
            }
        }
        if hasWin { return true }
    }
    return false
}


type TicTacToe struct {
  board *Board
  currentPlayer int
  ongoingGame bool
  remainingRounds int
}

func (ttt *TicTacToe) Init() *TicTacToe {
  // Create 3x3 board
  // currentPlayer starts at 'X' (0 index in board.symbol)
  ttt.board = new(Board).Init(3,3)
  ttt.currentPlayer = 0
  ttt.ongoingGame = true
  ttt.remainingRounds = 9
  return ttt
}

func (ttt *TicTacToe) play() {
    var row, col int
    for ttt.remainingRounds > 0 && ttt.ongoingGame {
      ttt.printCurrentPlayersTurn()
      ttt.board.printBoard()
      valid_move := false
      for !valid_move {
        if _, err := fmt.Scan(&row, &col); err != nil {
            ttt.printInvalidMove()
            continue
        } else {
            if !ttt.board.canPlacePiece(row, col) {
              ttt.printInvalidMove()
              continue
            }
            valid_move = true
            ttt.playRound(row, col)
        }
      }
    }
    if ttt.remainingRounds == 0 { ttt.declareTie() }
    ttt.board.printBoard()
}

func (ttt *TicTacToe) playRound(row int, col int) {
  ttt.board.placePiece(row, col, ttt.currentPlayer)
  ttt.printCurrentMove(row, col)
  if ttt.board.checkWinCondtion(ttt.currentPlayer) {
      ttt.declareWinner()
      ttt.ongoingGame = false
  } else {
      ttt.switchPlayer()
      ttt.decrementRounds()
  }
}

func (ttt *TicTacToe) getCurrPlayer() string{
    return string(ttt.board.symbol[ttt.currentPlayer])
}

func (ttt *TicTacToe) printCurrentPlayersTurn() {
  fmt.Println("This is Player " + ttt.getCurrPlayer() + "'s turn.")
}

func (ttt *TicTacToe) decrementRounds() {
  ttt.remainingRounds--
}

func (ttt *TicTacToe) declareWinner() {
  fmt.Println("Player " + ttt.getCurrPlayer() + " won!")
}

func (ttt *TicTacToe) declareTie() {
  fmt.Println("The game is tied")
}

func (ttt *TicTacToe) printInvalidMove(){
  fmt.Println("Please enter valid unmarked row (0 <= row <"+string(ttt.board.numRows)+" and column (0 <= column <"+string(ttt.board.numCols)+") (e.g. 1 0)\n")
}

func (ttt *TicTacToe) printCurrentMove(row int, col int) {
  fmt.Println("Player " + ttt.getCurrPlayer() + " made a move at row " + string(row) + " and column " + string(col) + "!\n")
}

func (ttt *TicTacToe) switchPlayer() {
  ttt.currentPlayer = 1 - ttt.currentPlayer
}


func main() {
  ttt := new(TicTacToe).Init()
  ttt.play()
}

File: target_practice_solutions/java/w4_d1_oop.java
import java.util.*;
import java.io.*;

class TTTRunner {
    public static void main(String[] args) {
        TicTacToe ttt = new TicTacToe();
        ttt.play();
    }
}

class TicTacToe {

    private Board board;
    private int currentPlayer;
    private boolean ongoing;
    private int remainingRounds;

    public TicTacToe() {
        // Current Player starts at X
        // Number of Rounds
        // Create 3x3 board
        board = new Board();
        currentPlayer = 0;
        remainingRounds = 9;
        ongoing = true;
    }

    public void play() {
        Scanner in = new Scanner(new BufferedReader(new InputStreamReader(System.in)));
        int row, col;
        while (remainingRounds > 0 && ongoing) {
            printCurrentPlayersTurn();
            board.printBoard();
            try {
                row = in.nextInt();
                col = in.nextInt();
                if (!board.canPlacePiece(row, col)) throw new InputMismatchException();
                playRound(row, col);
            } catch (InputMismatchException e){
                printInvalidMove();
            }
            in.nextLine();
        }
        if (remainingRounds == 0) declareTie();
        board.printBoard();
    }

    public void playRound(int row, int col) {
        board.placePiece(row, col, currentPlayer);
        printCurrentMove(row, col);
        if (board.checkWinCondtion(currentPlayer)) {
            declareWinner();
            ongoing = false;
        } else {
            switchPlayer();
            decrementRounds();
        }
    }

    public char getCurrPlayer(){
        return board.symbol[currentPlayer];
    }

    public void printCurrentPlayersTurn() {
        System.out.println("This is Player " + getCurrPlayer() + "'s turn.");
    }

    public void decrementRounds() {
        remainingRounds--;
    }

    public void declareWinner() {
        System.out.println("Player " + getCurrPlayer() + " won!");
    }

    public void declareTie() {
        System.out.println("The game is tied");
    }

    public void printInvalidMove(){
        System.out.println("Please enter valid unmarked row (0 <= row <3) and column (0 <= column <3) (e.g. 1 0)\n");
    }

    public void printCurrentMove(int row, int col) {
        System.out.println("Player " + getCurrPlayer() + " made a move at row " + row + " and column " + col + "!\n");
    }

    public void switchPlayer() {
        currentPlayer = 1 - currentPlayer;
    }

}

class Board {

    private Character[][] matrix;
    char[] symbol;

    public Board() {
        matrix = new Character[3][3];
        symbol = new char[]{'O', 'X'};
    }

    public void printBoard() {
        StringBuilder output = new StringBuilder();
        for (Character[] characters : matrix) {
            output.append("|");
            for (Character character : characters) {
                output.append(" ");
                output.append(character == null ? '-' : character);
                output.append(" ");
                output.append("|");
            }
            output.append("\n");
        }
        System.out.println(output.toString());
    }

    public boolean canPlacePiece(int row, int col) {
        return (row < 3 && row >= 0)
                && (col < 3 && col >= 0)
                && matrix[row][col] == null;
    }

    public void placePiece(int row, int col, int player) {
        if (canPlacePiece(row, col)) {
            matrix[row][col] = symbol[player];
        }
    }

    public boolean checkWinCondtion(int player) {
        return checkRows(player) || checkColumns(player) || checkDiagonals(player);
    }

    public boolean checkDiagonals(int player) {
        boolean hasWin = true;
        for (int i=0; i<matrix.length; i++) {
            if (matrix[i][i] == null || matrix[i][i] != symbol[player]) {
                hasWin = false;
                break;
            }
        }
        if (hasWin) return true;
        hasWin = true;
        for (int i=0; i<matrix.length; i++) {
            if (matrix[i][matrix.length-1 - i] == null || matrix[i][matrix.length-1 - i] != symbol[player]) {
                hasWin = false;
                break;
            }
        }
        return hasWin;
    }

    public boolean checkRows(int player) {
        boolean hasWin;
        for (int row=0; row<matrix.length; row++) {
            hasWin = true;
            for (int col=0; col<matrix[row].length; col++) {
                if (matrix[row][col] == null || matrix[row][col] != symbol[player]) {
                    hasWin = false;
                    break;
                }
            }
            if (hasWin) return true;
        }
        return false;
    }

    public boolean checkColumns(int player) {
        boolean hasWin;
        for (int col=0; col<matrix[0].length; col++) {
            hasWin = true;
            for (int row=0; row<matrix.length; row++) {
                if (matrix[row][col] == null || matrix[row][col] != symbol[player]) {
                    hasWin = false;
                    break;
                }
            }
            if (hasWin) return true;
        }
        return false;
    }
}
